File: leetcode/trees/uniqueBinarySearchTree.js
/**
 * @param {number} n
 * @return {number}
 - n = 1 return 1
 - n = 2 0 + 1 + 1, 1 + 1 + 0 - 2
    1      2
     \    /
      2  1
 - 3 - 0 + 1 + 2, 1 + 1 + 1, 2 + 1 + 0 = 5
           2    +     1     +    2
 - 4 - 0 + 1 + 3, 1 + 1 + 2, 2 + 1 + 1, 3 + 1 + 0
           5    +     2    +    2     +     5
             n
    - loop i = 0, n
    0 + 1 + n - 1, 1 + 1 + n - 2, ..., n - 1, 1, 0
             
 */
    var numTrees = function(n) {
        return recursiveCall(n + 1, {1: 1, 0: 1});
    };
    
    function recursiveCall(n, cache) {
        if(cache[n]) return cache[n];
        let sum = 0;
        for(let i = 1; i < n; i++) {
            sum += recursiveCall(i, cache) * recursiveCall(n - i, cache);
        }
        return cache[n] = sum;
    }
    

File: leetcode/trees/uniqueBinarySearchTree.go
/*

1-1
2-2
3-1+2, 1+1+1, 2+1 - 2 + 1 + 2 = 5
4-1+3, 1+1+2, 2+1+1, 3+1 = 5 + 2 + 2 + 5 = 14
3
0*2
1*1
2*1

4
0,1,2,3
0*3 - 5
1*2 - 2
2*1 - 2
3*0 - 5
*/
import "fmt"
func numTrees(n int) int {
    cache := make(map[int]int, 0)
    return recurseTree(n, cache)
}

func recurseTree(n int, cache map[int]int) int {
    if n == 0 || n == 1 {
        return 1
    }
    if val, ok := cache[n]; ok {
        return val
    }
    result := 0
    for i := 0; i < n; i++ {
        result += recurseTree(i, cache) * recurseTree(n - i - 1, cache)
    }
    cache[n] = result
    return result
}

